### PR TITLE
New List GoodieRole

### LINKED ISSUE #3135: New Role for list parsing
@mintsoft @moollaza We have quite a few Goodies that parse lists of items, I think having a Role that can deal with parsing various formats of lists (`[1, 2, 3]`, `1 and 2, and 3`, `1 2 3` etc.) as well as giving a way to format the lists in a standard way would be quite useful.

This Role would also be quite useful to be *used* in some of the other roles; allowing easy parsing of lists of dates or numbers.

@duckduckgo/community-leaders What do you think? If you guys agree I'll open a PR for this - I've been playing around with list parsing a bit with [shuffle](https://duck.co/ia/view/shuffle) so I've got some ideas for which libraries would be useful.

### PR DESCRIPTION
**Pull request by GuiltyDolphin**
###### Description of changes

Adds the 'Parse::List' Role for parsing and formatting lists.

###### Which issues (if any) does this fix?

Resolves #3135.

###### People to notify (@mention interested parties)

@mintsoft 

### TO-DO:

- [x] Add more options for formatting.
  - [ ] Perhaps use information (or allow the passing of.) from previous parse where appropriate for brackets/format.

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by russellholt** [comment]
Why do you want to format lists in perl? Why not do this in javascript, on the front end?

**Comment by GuiltyDolphin** [comment]
@russellholt Format as in *textually*, like NumberStyle formatting numbers, or Dates formatting dates in a standard manner.

I have no intention of rendering interactive lists :wink:

Anyway, that's not the main point of this - the main point is parsing.

**Comment by GuiltyDolphin** [comment]
@mintsoft Want to have a look over this? I may need to do some more fiddling later once we've got a bit of usage feedback.

File: lib/DDG/GoodieRole/Parse/List.pm [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    my @items = $options{nested} && %parens ? map {
+        is_list($_, %parens) ? parse_list($_, %options, %parens) : $_;
+    } @raw_items : @raw_items;
+    return unless all { $_ =~ /^$item$/ } @items;
+    return \@items;
+}
+
+sub format_list {
+    my $items = shift;
+    return "[@{[join ', ', @$items]}]";
```
Comment (mintsoft) [inline lib/DDG/GoodieRole/Parse/List.pm]: Personally I'm not a fan of that .... feature

File: lib/DDG/GoodieRole/Parse/List.pm [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    ($list_text, my %parens) = remove_parens($list_text);
+    return [] if $list_text eq '';
+    my $sep = get_separator($list_text);
+    my $parens = join '', @{$parens{parens} // []};
+    my $record = Data::Record->new({
+        split => $sep,
+        unless => $options{nested} && $parens ? qr/(?:$RE{quoted}|$RE{balanced}{-parens=>$parens})/ : $RE{quoted},
+    });
+    my @raw_items = map { trim_whitespace $_ } $record->records($list_text);
+    my @items = $options{nested} && %parens ? map {
```
Comment (mintsoft) [inline lib/DDG/GoodieRole/Parse/List.pm]: Can we break this out a bit into some variables?

File: lib/DDG/GoodieRole/Parse/List.pm (hunk 1 above)
Comment (GuiltyDolphin) [inline lib/DDG/GoodieRole/Parse/List.pm]: Uhh, whoops! I must've pushed to the wrong branch, this is *really* outdated.

File: lib/DDG/GoodieRole/Parse/List.pm (hunk 2 above)
Comment (GuiltyDolphin) [inline lib/DDG/GoodieRole/Parse/List.pm]: Which bit? `@items`?

**Comment by GuiltyDolphin** [comment]
@mintsoft Sorry about that - must've forgotten to push or something.

File: lib/DDG/GoodieRole/Parse/List.pm (hunk 2 above)
Comment (mintsoft) [inline lib/DDG/GoodieRole/Parse/List.pm]: Yeah; I think the double nested ternary and map is a little ... difficult to read

**Comment by GuiltyDolphin** [comment]
@mintsoft That better?

**Comment by GuiltyDolphin** [comment]
@mintsoft Anything else? Any other features you'd like to see before a first release? I'm thinking we only really need `format_list` and `parse_list` though.

**Comment by mintsoft** [comment]
Yeah I think basic parsing and formatting is enough; we can always add new stuff if we see a need

**Comment by GuiltyDolphin** [comment]
@moollaza @mintsoft Le pingy ping. This should be good to go unless there's any immediate changes you want to see?

**Comment by GuiltyDolphin** [comment]
@mintsoft @moollaza Yo, this good to go? It's been sitting here a while.

**Comment by mintsoft** [comment]
@GuiltyDolphin LGTM

**Comment by mintsoft** [comment]
@GuiltyDolphin @moollaza I'm going to merge this in now as it has no effect on any production code and the like. We could probably do with updating the docs though to make it obvious that it's available for use

**Pull request merged**